OpenOversight/app/email_client.py

```diff
@@ -1,56 +1,160 @@
-from apiclient import errors
+import os.path
+from abc import ABC, abstractmethod
+from threading import Thread
+from typing import Optional
+
 from flask import current_app
+from flask_mail import Mail, Message
 from google.oauth2 import service_account
 from googleapiclient.discovery import build
 
 from OpenOversight.app.models.emails import Email
 from OpenOversight.app.utils.constants import KEY_OO_SERVICE_EMAIL, SERVICE_ACCOUNT_FILE
 
 
-class EmailClient(object):
+class EmailProvider(ABC):
+    """Base class to define how emails are sent."""
+
+    @abstractmethod
+    def start(self):
+        """Set up the email provider."""
+
+    @abstractmethod
+    def is_configured(self):
+        """Determine the required env variables for this provider are configured."""
+
+    @abstractmethod
+    def send_email(self, email: Email):
+        """Send an email with this email provider."""
+
+
+class GmailEmailProvider(EmailProvider):
+    """Sends email through Gmail using the Google API client."""
+
+    SCOPES = ["https://www.googleapis.com/auth/gmail.send"]
+
+    def start(self):
+        credentials = service_account.Credentials.from_service_account_file(
+            SERVICE_ACCOUNT_FILE, scopes=self.SCOPES
+        )
+        delegated_credentials = credentials.with_subject(
+            current_app.config[KEY_OO_SERVICE_EMAIL]
+        )
+        self.service = build("gmail", "v1", credentials=delegated_credentials)
+
+    def is_configured(self):
+        return (
+            os.path.isfile(SERVICE_ACCOUNT_FILE)
+            and os.path.getsize(SERVICE_ACCOUNT_FILE) > 0
+        )
+
+    def send_email(self, email: Email):
+        (
+            self.service.users()
+            .messages()
+            .send(userId="me", body=email.create_message())
+            .execute()
+        )
+
+
+class SMTPEmailProvider(EmailProvider):
+    """Sends email with SMTP using Flask-Mail."""
+
+    CONFIG_SENDER_KEY = "OO_MAIL_SENDER"
```

Comment:
I don't see a `OO_SERVICE_EMAIL` value in the config. These two are options: `OO_SERVICE_EMAIL` and `OO_HELP_EMAIL`. I'd be hesitant to add another set of configuration values unless there is an absolute need.

Reply:
Oops, this was directly pulled from the old email code. I've removed it from `is_configured` since it's used for both Gmail and SMTP and replaced it with `OO_SERVICE_EMAIL` in the email sending code.
https://github.com/lucyparsons/OpenOversight/commit/c964a94c948d7aa910702277af9cfed62684c83b#diff-b80bc83f4119bbc14d3e988237671625aa481e5904a92992bb786bcb0303f3deL14-L20

Reply:
I'm a bit confused. I still see `OO_MAIL_SENDER` in the code. All instances of it should have been replaced with `OO_SERVICE_EMAIL`, yeah?

Reply:
It should be replaced now